pkg/sdk/reconciler.go

```diff
@@ -0,0 +1,83 @@
+package sdk
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctlr "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type reconciler[T ObjectConstraint] struct {
+	client.Client
+	impl         Implementation[T]
+	resourceKind string
+}
+
+var _ reconcile.Reconciler = &reconciler[client.Object]{}
+
+func RegisterController[T ObjectConstraint](mgr manager.Manager, impl Implementation[T]) error {
+	return RegisterControllerWithEventFilter(mgr, impl, nil)
+}
+
+// To handle new controllers from PR https://github.com/nginxinc/nginx-kubernetes-gateway/pull/221
+func RegisterControllerWithEventFilter[T ObjectConstraint](
+	mgr manager.Manager,
+	impl Implementation[T],
+	filter predicate.Predicate,
+) error {
+	var obj T
+
+	r := &reconciler[T]{
+		Client:       mgr.GetClient(),
+		impl:         impl,
+		resourceKind: reflect.TypeOf(obj).Elem().Name(),
+	}
+
+	obj = reflect.New(reflect.TypeOf(obj).Elem()).Interface().(T)
```

Comment:
Did you try the builtin `new`? The interaction between template functions and their non-template counterparts is interesting.

```
var obj *T
obj = new(T)
builder := ctlr.NewControllerManagedBy(mgr).For(*obj)
```

A buildable experiment:
```
package main

import (
	"fmt"
	"reflect"
	"time"
)

type foo struct {
	t time.Time
}

type bar struct {
	t time.Time
}

type timer interface {
	CreatedAt() time.Time
}

func (f *foo) CreatedAt() time.Time {
	return f.t
}

func (b *bar) CreatedAt() time.Time {
	return b.t
}

func NewFoo(t time.Time) *foo {
	return &foo{
		t: t,
	}
}

func NewBar(t time.Time) *bar {
	return &bar{
		t: t,
	}
}

type ObjectConstraint interface {
	timer
}

func checkCreated[T ObjectConstraint](check T) {
	var obj T
	fmt.Printf("TypeOf: %v\n", reflect.TypeOf(obj))
	fmt.Printf("TypeOf.Elem: %v\n", reflect.TypeOf(obj).Elem())
	fmt.Printf("TypeOf.Elem.Name: %v\n", reflect.TypeOf(obj).Elem().Name())
	
	fmt.Printf("New(TypeOf.Elem).Interface: %v\n", 
		reflect.New(reflect.TypeOf(obj).Elem()).Interface())
	fmt.Println()
	fmt.Println()

	obj = reflect.New(reflect.TypeOf(obj).Elem()).Interface().(T)
  fmt.Printf("TypeOf: %v\n", reflect.TypeOf(obj))
  fmt.Printf("TypeOf.Elem: %v\n", reflect.TypeOf(obj).Elem())
  fmt.Printf("TypeOf.Elem.Name: %v\n", reflect.TypeOf(obj).Elem().Name())

  fmt.Printf("New(TypeOf.Elem).Interface: %v\n",
    reflect.New(reflect.TypeOf(obj).Elem()).Interface())
	fmt.Println()
	fmt.Println()

	obj2 := new(T)
  fmt.Printf("TypeOf: %v\n", reflect.TypeOf(*obj2))
  fmt.Printf("TypeOf.Elem: %v\n", reflect.TypeOf(*obj2).Elem())
  fmt.Printf("TypeOf.Elem.Name: %v\n", reflect.TypeOf(*obj2).Elem().Name())

  fmt.Printf("New(TypeOf.Elem).Interface: %v\n",
    reflect.New(reflect.TypeOf(*obj2).Elem()).Interface())

	fmt.Println(check.CreatedAt())
}

func main() {
	f := NewFoo(time.Now())
	//b := NewBar(time.Now())

	checkCreated(f)
	//checkCreated(b)
}
```

Reply:
> Did you try the builtin new? The interaction between template functions and their non-template counterparts is interesting.

```
var obj *T
obj = new(T)
builder := ctlr.NewControllerManagedBy(mgr).For(*obj)
```

In this case, *obj will be pointer to a nil value. This in turn will make the builder fail on the `Complete()` or `Build()` call because of the nil check there.

Reply:
```
	obj2 := new(T)
	timerRef := *obj2
	fmt.Printf("Value: %v\n", timerRef)
```
leads to
```
Value: <nil>
```